### PR TITLE
Full build with Versioning, GH Releases and Artifacts

### PR DESCRIPTION
**Pull request by szlatkow**

### CODE REVIEW

File: .github/workflows/ci-cd.yml [hunk 1]
```diff
@@ -54,42 +54,80 @@ jobs:
...
+          $gitNewVersion = if ("${{ steps.tag_generator.outputs.new_version }}") {"${{ steps.tag_generator.outputs.new_version }}"} else {$null}
+          $projectCurrentVersion = ([xml](Get-Content .\src\GuardClauses\GuardClauses.csproj)).Project.PropertyGroup.Version
+          $projectNewVersion = $gitNewVersion ?? $projectCurrentVersion
+          dotnet build .\GuardClauses.sln --configuration Release -property:Version=$projectNewVersion
 
       - name: Run Unit Tests
         id: run_unit_tests
         shell: pwsh
         run: |
           dotnet test .\GuardClauses.sln --verbosity normal
```
Comment (aleks-ivanov) [inline .github/workflows/ci-cd.yml]: Consider --no-restore/--no-build flags and a --configuration property. We will discuss this tomorrow.

File: .github/workflows/ci-cd.yml [hunk 2]
```diff
@@ -54,42 +54,80 @@ jobs:
...
+        with:
+          PROJECT_FILE_PATH: src/GuardClauses/GuardClauses.csproj
+          VERSION_REGEX: <Version>(.*)<\/Version>
+          TAG_COMMIT: true
+          TAG_FORMAT: GuardClauses-v*
+          NUGET_KEY: ${{secrets.NUGET_API_KEY}}
+
+      - name: Create and publish release
+        id: create_release
+        uses: softprops/action-gh-release@v1
```
Comment (aleks-ivanov) [inline .github/workflows/ci-cd.yml:134]: Consider adding more attributes if any could be used to reproduce the way the owner manually submits their release. We will discuss this tomorrow.

File: .github/workflows/ci-cd.yml [hunk 3]
```diff
@@ -54,42 +54,80 @@ jobs:
...
 
+      - if: steps.conditionals_handler.outputs.is_push_to_default_branch == 'true'
+        name: Bump GH tag
+        id: tag_generator
+        uses: mathieudutour/github-tag-action@v5.4
+        with: 
+          github_token: ${{ secrets.GITHUB_TOKEN }}
+          default_bump: false
+          release_branches: ${{ steps.data_gatherer.outputs.default_branch }}
+
```
Comment (aleks-ivanov) [inline .github/workflows/ci-cd.yml:72]: An important step is missing! Pay closer attention to the already existing pipelines. We will discuss this tomorrow.

**Review by aleks-ivanov — CHANGES_REQUESTED** [review]

File: .github/workflows/ci-cd.yml (hunk 2 above)
Comment (szlatkow) [reply .github/workflows/ci-cd.yml:134]: When you add a comment under the prefix for the release the Release is taking the commit message and publishing it like the owner is publishing his release.

File: .github/workflows/ci-cd.yml (hunk 3 above)
Comment (szlatkow) [reply .github/workflows/ci-cd.yml:72]: > An important step is missing! Pay closer attention to the already existing pipelines. We will discuss this tomorrow.

Resolved

**Review by aleks-ivanov — DISMISSED** [review]

File: .github/workflows/ci-cd.yml [hunk 4]
```diff
@@ -41,7 +42,6 @@ jobs:
           Write-Output "::set-output name=is_push::$(echo $isPush)"
           Write-Output "::set-output name=is_push_to_default_branch::$(echo $isPushToDefaultBranch)"
 
-
```
Comment (aleks-ivanov) [inline .github/workflows/ci-cd.yml:44]: test

**Review by aleks-ivanov — CHANGES_REQUESTED** [review]

File: .github/workflows/ci-cd.yml (hunk 4 above)
Comment (szlatkow) [reply .github/workflows/ci-cd.yml:44]: test

**Review by szlatkow — COMMENTED** [review]
test

File: .github/workflows/ci-cd.yml [hunk 5]
```diff
@@ -112,7 +112,16 @@ jobs:
           name: GuardClauses v${{ needs.ci.outputs.latest_version }}
           path: Artifacts\
 
-      - name: Publish to NuGet
+      - name: Filter changes
```
Comment (aleks-ivanov) [inline .github/workflows/ci-cd.yml]: Use the logic from this action for the whole CD job

**Review by aleks-ivanov — CHANGES_REQUESTED** [review]

File: .github/workflows/ci-cd.yml (hunk 5 above)
Comment (szlatkow) [reply .github/workflows/ci-cd.yml]: Moved the filter changes step after the checkout in order to work

**Review by aleks-ivanov — APPROVED** [review]

**Comment by szlatkow** [comment]
@aleks-ivanov I am ready, you can merge the pull request.

**Pull request merged**